File: web/explorer/src/api/sessionHint.ts
'use client';

import { useSyncExternalStore } from 'react';

/**
 * The session hint — a JS-readable presence flag the API writes beside
 * the HttpOnly session cookie (`internal/api/v1/dashboardauth/session_hint.go`).
 *
 * The explorer and the API are separate origins, so the only way this
 * app could tell a signed-in visitor from an anonymous one was to make
 * a credentialed `GET /v1/account/me` and read the status back. For an
 * anonymous visitor that is a guaranteed 401 on every page load, plus a
 * refetch every five minutes — and the browser writes its own console
 * entry for any non-2xx, from the network layer, where no `catch` can
 * reach it. The hint lets an anonymous visitor make the request zero
 * times instead.
 *
 * Three properties matter:
 *
 *  - It is a HINT. Nothing is authorized by it; the session cookie is
 *    still the credential, still HttpOnly, still checked server-side on
 *    every request. Forging this cookie buys an attacker one 401.
 *  - Its absence is treated as "signed out", which is also what an
 *    older API that doesn't set it yet produces. That degrades to the
 *    signed-out CTAs — never to a broken signed-in state.
 *  - It can outlive the real session (server-side expiry, revocation,
 *    one side cleared without the other). The probe then runs and 401s
 *    exactly as it does today, and `useMe` drops the stale hint so the
 *    next page load is quiet again.
 */
export const SESSION_HINT_COOKIE = 'stellarindex_session_present';

/** Subscribers to local hint changes (i.e. clearSessionHint). */
const listeners = new Set<() => void>();

function subscribe(onStoreChange: () => void): () => void {
  listeners.add(onStoreChange);
  return () => {
    listeners.delete(onStoreChange);
  };
}

/**
 * Reads one cookie out of a `document.cookie` jar string. An empty
 * value reads as absent: a cookie cleared with `name=` rather than an
 * expiry would otherwise look like a live session.
 */
export function readCookie(jar: string, name: string): string | null {
  for (const part of jar.split(';')) {
    const eq = part.indexOf('=');
    if (eq < 0) continue;
    if (part.slice(0, eq).trim() !== name) continue;
    const value = part.slice(eq + 1).trim();
    return value === '' ? null : value;
  }
  return null;
}

/** Whether this browser is holding a session hint right now. */
export function sessionHintPresent(): boolean {
  if (typeof document === 'undefined') return false;
  return readCookie(document.cookie, SESSION_HINT_COOKIE) !== null;
}

/**
 * The domain attributes a deletion has to be attempted with. Removing a
 * cookie requires the same Domain and Path it was set with, and this
 * app does not know the API's configured `cookie_domain` — so try the
 * host-only form and then every parent up to (not past) the registrable
 * pair. The attempts that don't match are inert, and a browser rejects
 * outright any that names a public suffix.
 */
function candidateDomains(hostname: string): (string | undefined)[] {
  const candidates: (string | undefined)[] = [undefined];
  const labels = hostname.split('.');
  for (let i = 0; i <= labels.length - 2; i++) {
    candidates.push(labels.slice(i).join('.'));
  }
  return candidates;
}

/**
 * Drops a hint that has outlived its session, so the next page load
 * makes no credentialed request at all.
 */
export function clearSessionHint(): void {
  if (typeof document === 'undefined') return;
  const secure = window.location.protocol === 'https:' ? '; Secure' : '';
  for (const domain of candidateDomains(window.location.hostname)) {
    document.cookie =
      `${SESSION_HINT_COOKIE}=; Max-Age=0; Path=/; SameSite=Lax` +
      (domain ? `; Domain=${domain}` : '') +
      secure;
  }
  for (const listener of listeners) listener();
}

/** Server/prerender snapshot — a static export has no cookie jar. */
function absent(): boolean {
  return false;
}

/**
 * useSessionHint — re-renders when the hint is dropped, so a query
 * gated on it stops refetching the moment it goes stale rather than at
 * the next accidental render.
 */
export function useSessionHint(): boolean {
  return useSyncExternalStore(subscribe, sessionHintPresent, absent);
}

File: internal/api/v1/dashboardauth/session_hint.go
package dashboardauth

import (
	"net/http"
	"time"
)

// SessionHintCookieName is a JS-readable presence flag that shadows
// [SessionCookieName]. It is written and cleared in lockstep with the
// session cookie and carries nothing else.
//
// Why it exists: the explorer is a separate origin from the API, so the
// only way it could tell a signed-in visitor from an anonymous one was
// to make a credentialed `GET /v1/account/me` and read the status. For
// an anonymous visitor that is a guaranteed 401 on every page load
// (plus a refetch every five minutes), and a browser emits its own
// console entry for any non-2xx response — an entry no amount of
// client-side error handling can suppress, because it comes from the
// network layer rather than from the promise.
//
// The session cookie itself cannot answer the question: it is HttpOnly
// (and must stay that way — it is the bearer credential). So the same
// responses that write the session cookie also write this one, which is
// NOT HttpOnly. `document.cookie` sees it, and the explorer skips the
// probe entirely when it is absent.
//
// It is a HINT, never an authorization signal:
//
//   - It carries no identity, no token and no email — the value is the
//     single byte "1". Anything an attacker could forge, they could
//     equally forge by setting this cookie themselves, and the worst
//     that buys them is one 401.
//   - Nothing server-side reads it. Every authenticated route still
//     resolves [SessionCookieName] against the sessions table.
//   - It can outlive the real session (server-side expiry, revocation,
//     one cookie cleared without the other). That is expected: the
//     probe then runs and 401s exactly as it does today, and the client
//     drops the stale hint.
const SessionHintCookieName = "stellarindex_session_present"

// sessionHintValue is the entire contents of the hint cookie. A
// constant, so no per-user information can reach it by accident.
const sessionHintValue = "1"

// setSessionHintCookie writes the presence flag alongside a freshly
// minted session cookie. Every attribute except HttpOnly mirrors the
// session cookie so the pair shares one scope and one lifetime.
func (h *Handlers) setSessionHintCookie(w http.ResponseWriter, expires time.Time) {
	http.SetCookie(w, &http.Cookie{
		Name:    SessionHintCookieName,
		Value:   sessionHintValue,
		Path:    "/",
		Domain:  h.cfg.CookieDomain,
		Expires: expires,
		// Deliberately readable from JavaScript — that is the whole
		// point of this cookie. It is safe because the value is a
		// constant with no bearer power (see [SessionHintCookieName]).
		HttpOnly: false,
		Secure:   h.cfg.CookieSecure,
		SameSite: sessionSameSite(),
	})
}

// clearSessionHintCookie drops the presence flag. Called from the same
// places that clear the session cookie, so a browser can never be left
// holding a hint for a session the server has already dropped on its
// own initiative.
func (h *Handlers) clearSessionHintCookie(w http.ResponseWriter) {
	http.SetCookie(w, &http.Cookie{
		Name:     SessionHintCookieName,
		Value:    "",
		Path:     "/",
		Domain:   h.cfg.CookieDomain,
		MaxAge:   -1,
		HttpOnly: false,
		Secure:   h.cfg.CookieSecure,
		SameSite: sessionSameSite(),
	})
}
